source/extensions/filters/http/oauth2/filter.cc

```diff
@@ -491,61 +498,62 @@ void OAuth2Filter::redirectToOAuthServer(Http::RequestHeaderMap& headers) const
+      return key == config_->cookieNames().oauth_nonce_;
+    });
+    if (nonce_cookie.find(config_->cookieNames().oauth_nonce_) != nonce_cookie.end()) {
+      nonce = nonce_cookie.at(config_->cookieNames().oauth_nonce_);
+      nonce_cookie_exists = true;
+    } else {
+      nonce = generateNonce(config_->hmacSecret(), time_source_);
+    }
 
-  // Set the nonce cookie if it does not exist.
-  if (!nonce_cookie_exists) {
-    // Expire the nonce cookie in 10 minutes.
-    // This should be enough time for the user to complete the OAuth flow.
-    std::string expire_in = std::to_string(10 * 60);
-    std::string cookie_tail_http_only = fmt::format(CookieTailHttpOnlyFormatString, expire_in);
-    if (!config_->cookieDomain().empty()) {
-      cookie_tail_http_only = absl::StrCat(
-          fmt::format(CookieDomainFormatString, config_->cookieDomain()), cookie_tail_http_only);
+    // Set the nonce cookie if it does not exist.
+    if (!nonce_cookie_exists) {
+      // Expire the nonce cookie in 10 minutes.
+      // This should be enough time for the user to complete the OAuth flow.
+      std::string expire_in = std::to_string(10 * 60);
+      std::string cookie_tail_http_only = fmt::format(CookieTailHttpOnlyFormatString, expire_in);
+      if (!config_->cookieDomain().empty()) {
+        cookie_tail_http_only = absl::StrCat(
+            fmt::format(CookieDomainFormatString, config_->cookieDomain()), cookie_tail_http_only);
+      }
+      response_headers->addReferenceKey(
+          Http::Headers::get().SetCookie,
+          absl::StrCat(config_->cookieNames().oauth_nonce_, "=", nonce, cookie_tail_http_only));
     }
-    response_headers->addReferenceKey(
-        Http::Headers::get().SetCookie,
-        absl::StrCat(config_->cookieNames().oauth_nonce_, "=", nonce, cookie_tail_http_only));
+    state = encodeState(original_url, nonce);
+  } else {
+    state = Http::Utility::PercentEncoding::urlEncodeQueryParameter(original_url);
   }
 
-  // Encode the original request URL and the nonce to the state parameter
-  const std::string state =
-      absl::StrCat(stateParamsUrl, "=", escaped_url, "&", stateParamsNonce, "=", nonce);
-  const std::string escaped_state = Http::Utility::PercentEncoding::urlEncodeQueryParameter(state);
+  auto query_params = config_->authorizationQueryParams();
+  query_params.overwrite(queryParamsState, state);
 
   Formatter::FormatterPtr formatter = THROW_OR_RETURN_VALUE(
       Formatter::FormatterImpl::create(config_->redirectUri()), Formatter::FormatterPtr);
   const auto redirect_uri =
       formatter->formatWithContext({&headers}, decoder_callbacks_->streamInfo());
   const std::string escaped_redirect_uri =
       Http::Utility::PercentEncoding::urlEncodeQueryParameter(redirect_uri);
-
-  auto query_params = config_->authorizationQueryParams();
   query_params.overwrite(queryParamsRedirectUri, escaped_redirect_uri);
-  query_params.overwrite(queryParamsState, escaped_state);
+
```

Comment:
From the title and PR description, I think what we expect in the PR is only change the way to encode the state to avoid AWS cognito complaining.

Disable nonce should be another different topic.